# Preparation

In [1]:
import pandas as pd
import numpy as np

In [4]:
frame = pd.read_csv('rating_final.csv')
cuisine = pd.read_csv('chefmozcuisine.csv')
geodata = pd.read_csv('geoplaces2.csv', encoding='latin-1')

In [6]:
frame.head()

userID  placeID  rating  food_rating  service_rating
0  U1077   135085       2            2               2
1  U1077   135038       2            2               1
2  U1077   132825       2            2               2
3  U1077   135060       1            2               2
4  U1068   135104       1            1               2

In [7]:
cuisine.head()

placeID        Rcuisine
0   135110         Spanish
1   135109         Italian
2   135107  Latin_American
3   135106         Mexican
4   135105       Fast_Food

In [15]:
geodata.head()

placeID   latitude   longitude  \
0   134999  18.915421  -99.184871   
1   132825  22.147392 -100.983092   
2   135106  22.149709 -100.976093   
3   132667  23.752697  -99.163359   
4   132613  23.752903  -99.165076   

                                      the_geom_meter  \
0  0101000020957F000088568DE356715AC138C0A525FC46...   
1  0101000020957F00001AD016568C4858C1243261274BA5...   
2  0101000020957F0000649D6F21634858C119AE9BF528A3...   
3  0101000020957F00005D67BCDDED8157C1222A2DC8D84D...   
4  0101000020957F00008EBA2D06DC8157C194E03B7B504E...   

                             name                                 address  \
0                 Kiku Cuernavaca                              Revolucion   
1                 puesto de tacos  esquina santos degollado y leon guzman   
2      El Rincón de San Francisco                         Universidad 169   
3  little pizza Emilio Portes Gil                 calle emilio portes gil   
4                   carnitas_mata                  lic. Emilio portes gil   

              city            state country fax      ...        \
0       Cuernavaca          Morelos  Mexico   ?      ...         
1           s.l.p.           s.l.p.  mexico   ?      ...         
2  San Luis Potosi  San Luis Potosi  Mexico   ?      ...         
3        victoria        tamaulipas       ?   ?      ...         
4         victoria       Tamaulipas  Mexico   ?      ...         

             alcohol smoking_area dress_code     accessibility   price  \
0  No_Alcohol_Served         none   informal  no_accessibility  medium   
1  No_Alcohol_Served         none   informal        completely     low   
2          Wine-Beer  only at bar   informal         partially  medium   
3  No_Alcohol_Served         none   informal        completely     low   
4  No_Alcohol_Served    permitted   informal        completely  medium   

                     url Rambience franchise    area other_services  
0  kikucuernavaca.com.mx  familiar         f  closed           none  
1                      ?  familiar         f    open           none  
2                      ?  familiar         f    open           none  
3                      ?  familiar         t  closed           none  
4                      ?  familiar         t  closed           none  

[5 rows x 21 columns]

geodata contains too much info - Extract from geodata only placeID and name

In [14]:
places = geodata[['placeID','name']]
places.head()

placeID                            name
0   134999                 Kiku Cuernavaca
1   132825                 puesto de tacos
2   135106      El Rincón de San Francisco
3   132667  little pizza Emilio Portes Gil
4   132613                   carnitas_mata

# Grouping and Ranking Data

In [8]:
#Create a new dataframe to aggregate the means of ratings for each PlaceID
rating = pd.DataFrame(frame.groupby('placeID')['rating'].mean())
rating.head()

rating
placeID        
132560     0.50
132561     0.75
132564     1.25
132572     1.00
132583     1.00

In [9]:
#Add a new column of 'rating' dataframe for popularity (count of ratings)
rating['rating_count'] = pd.DataFrame(frame.groupby('placeID')['rating'].count())
rating.head()

rating  rating_count
placeID                      
132560     0.50             4
132561     0.75             4
132564     1.25             4
132572     1.00            15
132583     1.00             4

In [10]:
rating.describe()

rating  rating_count
count  130.000000    130.000000
mean     1.179622      8.930769
std      0.349354      6.124279
min      0.250000      3.000000
25%      1.000000      5.000000
50%      1.181818      7.000000
75%      1.400000     11.000000
max      2.000000     36.000000

Interpretation: count = 130 indicates there're 130 unique places that have been reviewed in the rating data frame; max for rating_count = 36 indicates the most popular (most rated) place got a total of 36 reviews.

In [12]:
rating.sort_values('rating_count',ascending=False).head()

rating  rating_count
placeID                        
135085   1.333333            36
132825   1.281250            32
135032   1.178571            28
135052   1.280000            25
132834   1.000000            25

In [21]:
#Observe the name and cuisine type for placeID=135085
print(places[places['placeID']==135085])
print(cuisine[cuisine['placeID']==135085])

     placeID                    name
121   135085  Tortas Locas Hipocampo
    placeID   Rcuisine
44   135085  Fast_Food


# Preparing data for analysis

In [23]:
#Create pivot table for user and place, to observe matching directly
places_crosstab = pd.pivot_table(data=frame, values='rating', index='userID', columns='placeID')
places_crosstab.head()
#Note that places_crosstab is a dataframe

placeID  132560  132561  132564  132572  132583  132584  132594  132608  \
userID                                                                    
U1001       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
U1002       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
U1003       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
U1004       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
U1005       NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

placeID  132609  132613   ...    135080  135081  135082  135085  135086  \
userID                    ...                                             
U1001       NaN     NaN   ...       NaN     NaN     NaN     0.0     NaN   
U1002       NaN     NaN   ...       NaN     NaN     NaN     1.0     NaN   
U1003       NaN     NaN   ...       2.0     NaN     NaN     NaN     NaN   
U1004       NaN     NaN   ...       NaN     NaN     NaN     NaN     NaN   
U1005       NaN     NaN   ...       NaN     NaN     NaN     NaN     NaN   

placeID  135088  135104  135106  135108  135109  
userID                                           
U1001       NaN     NaN     NaN     NaN     NaN  
U1002       NaN     NaN     1.0     NaN     NaN  
U1003       NaN     NaN     NaN     NaN     NaN  
U1004       NaN     NaN     2.0     NaN     NaN  
U1005       NaN     NaN     NaN     NaN     NaN  

[5 rows x 130 columns]

In [24]:
#From the pivot table, extract all ratings for a restaurant called "Tortas"(with 36 ratings) as a series
Tortas_ratings = places_crosstab[135085]
Tortas_ratings.describe()

count    36.000000
mean      1.333333
std       0.755929
min       0.000000
25%       1.000000
50%       1.500000
75%       2.000000
max       2.000000
Name: 135085, dtype: float64

# Evaluating Similarity Based on Correlation

In [31]:
#Generate the Pearson R correlation coefficients between Tortas and each other places that's been reviewed
similar_to_Tortas = places_crosstab.corrwith(Tortas_ratings) #Will be an array for all Pearson R coefficients for each place

#Convert similar_to_Tortas matrix to dataframe
corr_Tortas = pd.DataFrame(similar_to_Tortas, columns=['PearsonR'])
corr_Tortas.dropna(inplace=True) #inplace = True means we don't need to assign a new data frame for it. 
corr_Tortas.head()

C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\numpy\lib\function_base.py:3175: RuntimeWarning: Degrees of freedom <= 0 for slice
  c = cov(x, y, rowvar)
C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\numpy\lib\function_base.py:3109: RuntimeWarning: divide by zero encountered in double_scalars
  c *= 1. / np.float64(fact)


PearsonR
placeID          
132572  -0.428571
132723   0.301511
132754   0.930261
132825   0.700745
132834   0.814823

From here, we need to take stock of how popular each of these places is as well. Simply observing on PearsonR may give us results which has similar rating with Tortas but only a few ratings. So we merge number of ratings into this corr_Tortas dataframe.

In [33]:
Tortas_corr_summary = corr_Tortas.join(rating['rating_count'])

#We want also places with ratings greater or equal to 10, so we create a filter and sort the counts
Tortas_corr_summary[Tortas_corr_summary['rating_count'] >= 10].sort_values('PearsonR', ascending=False).head(10)

PearsonR  rating_count
placeID                        
135076   1.000000            13
135085   1.000000            36
135066   1.000000            12
132754   0.930261            13
135045   0.912871            13
135062   0.898933            21
135028   0.892218            15
135042   0.881409            20
135046   0.867722            11
132872   0.840168            12

For places with "PearsonR = 1": only one user gave a review on both places, same score. But a correlation that's based on similarities between only one review rating is not meaningful. The places need to have more than one reviewer in common so we'll throw those places out.

Now let's observe the top correlated places, especially we might be interested in their cuisine type.

In [38]:
#Select the top 7 except for those thrown out
places_corr_Tortas = pd.DataFrame([135085, 132754, 135045, 135062, 135028, 135042, 135046], index = np.arange(7), columns=['placeID'])

summary = pd.merge(places_corr_Tortas, cuisine, on='placeID')
summary

placeID   Rcuisine
0   135085  Fast_Food
1   132754    Mexican
2   135028    Mexican
3   135042    Chinese
4   135046  Fast_Food

Only 5 places were returned because not all of the 7 top correlated places are in cuisine data frame, 2 were dropped when creating merged table.

Finally let's just observe the placeID = 135046 which has same cuisine type with Tortas

In [39]:
places[places['placeID']==135046]

placeID                     name
42   135046  Restaurante El Reyecito

This is most likely to be the place recommended to users who likes Tortas!